File: two-pointers/valid-palindrome.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
  public:
    bool isPalindrome(string s) {
      int i = 0, j = s.size() - 1;

      while (j > i) {
        while (!isalnum(s[i]) && j > i)
          ++i;

        while (!isalnum(s[j]) && j > i)
          --j;

        if (tolower(s[i++]) != tolower(s[j--]))
          return false;
      }

      return true;
    }
};

File: two-pointers/valid_palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        if not s:
            return True

        left, right = 0, len(s) - 1

        while left < right:
            while not s[left].isalnum() and left < right:
                left += 1
            while not s[right].isalnum() and left < right:
                right -= 1

            if s[left].lower() != s[right].lower():
                return False
            
            left += 1
            right -= 1

        return True